src/table/use-table-focus-navigation.ts

```diff
@@ -0,0 +1,166 @@
+      if (!currentFocusCell.current) {
+        return;
+      }
+      const [rowIndex, columnIndex] = currentFocusCell.current.slice();
+      let newRowIndex = rowIndex;
+      let newColumnIndex = columnIndex;
+
+      if (vertical !== 0) {
+        newRowIndex = Math.min(numRows, Math.max(rowIndex + vertical, 0));
+      }
+
+      if (horizontal !== 0) {
+        while (newColumnIndex <= maxColumnIndex && newColumnIndex >= minColumnIndex) {
+          newColumnIndex += horizontal;
+          if (focusableColumns[newColumnIndex]) {
+            break;
+          }
+        }
+      }
+
+      if (
+        (rowIndex !== newRowIndex || columnIndex !== newColumnIndex) &&
+        currentFocusCell.current &&
+        tableRoot.current
+      ) {
+        focusCell(newRowIndex, newColumnIndex);
+      }
+    },
+    [focusCell, focusableColumns, maxColumnIndex, minColumnIndex, numRows, tableRoot]
+  );
+
+  const handleArrowKeyEvents = useCallback(
+    (event: KeyboardEvent) => {
+      const abort =
+        !!tableRoot.current?.querySelector('[data-inline-editing-active = "true"]') ||
+        !document.activeElement?.closest('[data-inline-editing-active]');
+
+      if (abort) {
+        return;
+      }
+      switch (event.key) {
+        case 'ArrowUp':
+          event.preventDefault();
+          shiftFocus(-1, 0);
+          break;
+        case 'ArrowDown':
+          event.preventDefault();
+          shiftFocus(1, 0);
+          break;
+        case 'ArrowLeft':
+          event.preventDefault();
+          shiftFocus(0, -1);
+          break;
+        case 'ArrowRight':
+          event.preventDefault();
+          shiftFocus(0, 1);
+          break;
+        // istanbul ignore next (default case = do nothing, not testable)
+        default:
+          return;
+      }
+    },
+    [shiftFocus, tableRoot]
+  );
+
+  useEffect(() => {
+    const eventListeners = new Map<[number, number], { focusin(evt: any): void }>();
+    // istanbul ignore if
+    if (!tableRoot.current) {
+      return;
+    }
+
+    const tableElement = tableRoot.current;
+
+    // istanbul ignore next (tested in use-focus-navigation.test.tsx#L210)
+    function cleanUpListeners() {
+      iterateTableCells(tableElement, (cell, rowIndex, columnIndex) => {
+        const listeners = eventListeners.get([rowIndex, columnIndex]);
+        if (listeners?.focusin) {
+          cell.removeEventListener('focusin', listeners.focusin);
+        }
+      });
+      tableElement.removeEventListener('keydown', handleArrowKeyEvents);
+    }
+
+    iterateTableCells(tableElement, (cell, rowIndex, cellIndex) => {
+      if (!focusableColumns[cellIndex]) {
+        return;
+      }
+      const listenerFns = {
+        focusin: () => {
+          currentFocusCell.current = [rowIndex, cellIndex];
+        },
+      };
+      eventListeners.set([rowIndex, cellIndex], listenerFns);
+      cell.addEventListener('focusin', listenerFns.focusin, { passive: true });
+    });
```

Comment:
`focusin` event can bubble, so this can become a single listener

```js
function focusinCell(event) {
  const current = event.currentTarget;
  const td = findUpUntil(node => node.tagName === 'TD');
  if(!td) return;
  
  const tr = td.parentNode;
  const cellIndex = tr.children.indexOf(td);
  const rowIndex = tr.parentNode.indexOf(tr);

  currentFocusCell.current = [rowIndex, cellIndex];
}
```